jgrapht-core/src/main/java/org/jgrapht/alg/lca/TarjanLCAFinder.java

```diff
@@ -0,0 +1,202 @@
+    }
+
+    /**
+     * Construct a new instance of the algorithm.
+     *
+     * Note: If two roots appear in the same tree, an error will be thrown.
+     * Note: The constructor will NOT check if the input graph is a valid forest.
+     *
+     * @param graph the input graph
+     * @param roots the set of roots of the graph
+     */
+    public TarjanLCAFinder(Graph<V, E> graph, Set<V> roots) {
+        this.graph = Objects.requireNonNull(graph, "graph cannot be null");
+        this.roots = Objects.requireNonNull(roots, "roots cannot be null");
+
+        if (this.roots.isEmpty())
+            throw new IllegalArgumentException("roots cannot be empty");
+
+        if (!graph.vertexSet().containsAll(roots))
+            throw new IllegalArgumentException("at least one root is not a valid vertex");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public V getLCA(V a, V b) {
+        return getLCAs(Collections.singletonList(Pair.of(a,b))).get(0);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<V> getLCAs(List<Pair<V, V>> queries) {
+        return computeTarjan(queries);
+    }
+
+    private void initialize(){
+        unionFind = new UnionFind<>(Collections.emptySet());
+        ancestors = new HashMap<>();
+        blackNodes = new HashSet<>();
+    }
+
+    private void clear(){
+        unionFind = null;
+        ancestors = null;
+        blackNodes = null;
+        queryOccurs = null;
+
+        queries = null;
+        lowestCommonAncestors = null;
+    }
+
+    private List<V> computeTarjan(List<Pair<V, V>> queries){
+        initialize();
+
+        this.queries = queries;
+        this.lowestCommonAncestors = new ArrayList<>(queries.size());
+
+        this.queryOccurs = new HashMap<>();
+
+        for (int i = 0; i < queries.size(); i++){
+            V a = this.queries.get(i).getFirst();
+            V b = this.queries.get(i).getSecond();
+
+            if (!graph.vertexSet().contains(a))
+                throw new IllegalArgumentException("invalid vertex: " + a);
+
+            if (!graph.vertexSet().contains(b))
+                throw new IllegalArgumentException("invalid vertex: " + b);
+
+            if (a.equals(b))
+                this.lowestCommonAncestors.add(a);
+            else{
+                queryOccurs.computeIfAbsent(a, x -> new HashSet<>()).add(i);
+                queryOccurs.computeIfAbsent(b, x -> new HashSet<>()).add(i);
+
+                this.lowestCommonAncestors.add(null);
+            }
+        }
+
+        Set<V> visited = new HashSet<>();
+
+        for (V root: roots){
+            if (visited.contains(root))
+                throw new IllegalArgumentException("multiple roots in the same tree");
+
+            blackNodes.clear();
+            TarjanOLCA(root, null, visited);
+        }
+
+        List<V> tmpRef = lowestCommonAncestors;
+        clear();
+
+        return tmpRef;
+    }
+
+    private void TarjanOLCA(V u, V p, Set<V> visited){
```

Comment:
Method names start with a lowercase letter. I suggest to add a verb before the name (compute, use, find, try, etc.).